src/io/unix/iouring.rs

```diff
@@ -0,0 +1,176 @@
+            offset = offset + first_len;
+            self.write_all_at_aio(offset, second_buf, rio).await
+        } else {
+            self.sync
+                .write_append_all_buffers(first_buf, second_buf)
+                .await
+        }
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            self.write_all_at_aio(offset, buf, rio).await
+        } else {
+            self.sync.write_all_at(offset, buf).await
+        }
+    }
+
+    pub(crate) async fn read_all(&self) -> Result<BytesMut> {
+        self.read_exact_at_allocate(self.size().try_into()?, 0)
+            .await
+    }
+
+    pub(crate) async fn read_exact_at_allocate(
+        &self,
+        size: usize,
+        offset: u64,
+    ) -> Result<BytesMut> {
+        let buf = BytesMut::zeroed(size);
+        self.read_exact_at(buf, offset).await
+    }
+
+    pub(crate) async fn read_exact_at(&self, mut buf: BytesMut, offset: u64) -> Result<BytesMut> {
+        debug!("blob file read at");
+        if buf.is_empty() {
+            warn!("file read_at empty buf");
+        }
+        debug!(
+            "blob file read at buf len: {}, offset: {}",
+            buf.len(),
+            offset
+        );
+        if let Some(ref rio) = self.rio {
+            let compl = rio.read_at(&*self.sync.no_lock_fd, &buf, offset);
+            let mut size = compl.await.with_context(|| "read at failed")?;
+            while size < buf.len() {
+                debug!(
+                    "io uring read partial block, trying to read the rest, completed {}/{} bytes",
+                    size,
+                    buf.len()
+                );
+                let slice = buf.split_at_mut(size).1;
+                debug!("blob file read at buf to fill up: {}b", slice.len());
+                let compl = rio.read_at(&*self.sync.no_lock_fd, &slice, offset + size as u64);
+                let remainder_size = compl.await.with_context(|| "second read at failed")?;
+                debug!(
+                    "blob file read at second read, completed {}/{} of remains bytes",
+                    remainder_size,
+                    slice.len()
+                );
+                if remainder_size == 0 {
+                    let msg = "blob file read failed, second read returns zero bytes".to_string();
+                    return Err(Error::io(msg).into());
+                }
+                size += remainder_size;
+                debug!("blob file read at proggress {}/{}", size, buf.len());
+            }
+            debug!("blob file read at complited: {}", size);
+            Ok(buf)
+        } else {
+            self.sync.read_exact_at(buf, offset).await
+        }
+    }
+
+    pub(crate) async fn fsyncdata(&self) -> IOResult<()> {
+        if let Some(ref rio) = self.rio {
+            rio.fsync(&*self.sync.no_lock_fd).await
```

Comment:
You can add `fn std_file_ref(&self) -> &StdFile` function to the sync `File` and use it here and in other places